### ISSUE TITLE
Matrix types

### ISSUE THREAD

**Issue by vhirtham** [vhirtham/GDL#3]
**MatX**
- [x] Write Test for mat4X - first test that the single version is valid and use it afterwards to test if the SIMD versions are correct -> if a single version with 4 or more rows/cols is valid, all bigger matrices should be valid too.
- [x] (DROPPED) Check if the mat4XSIMD is faster if the register positions in the array are ordered row major (registers itself need to be column major) - this would actually make conversions between arrays of floats and the underlying matrix data more complicated and time consuming
- [x] Replace _mm_add_ps and _mm_mul_ps by _mm_fmadd_ps in mat4SIMD 
- [x] Create variadic template ctor for MatX
- [x] Make MatX available for more types than just floats/F32 - template signature should be the same as for matXSingle -> <type, rows, cols> CURRENT STATUS: Implement multiplication for size 2 registers!
- [x] Add transpose functions. Function without parameter returns a new object. ~~Function that takes a matrix as parameter transposes the object itself and returns a reference.~~ Static matrices can only be transosed in place if the number of registers won't change, which is not the case for all combinations of rows and columns
- [x] Use fixtures in test file
- [x] Add test for addition assignment
- [x] Check Data() function also in the multiplication test
- [x] Add multiplication test for two matrices with 6x2 and 2x6 size. (Tests a currently not tested case of issue #6)
- [x] add ~~constexpr~~(not possible for SSE) inline to classes. (inline to avoid multiple definitions issues) 
- [x] ~~Change Data function to deliver a pointer to the first element, like data() of std::containers~~ Not possible with MatXSIMD ---> Needs to create new array if `numValues % mNumRegisterEntries !=0`
- [x] ~~Try to use same header for Single and SIMD versions by using a template. Implementations are separated via template specialization. ---> Less code cause some stuff can be shared among all versions~~ Not possible for MatX since SIMD has some extra code in its header to adjust register sizes
- [x] use template meta programming in multiplication algorithm to handle different register sizes (http://www.cs.rpi.edu/~musser/design/blitz/meta-art.html)
- [x] ~~improve current static matX version by changing to dynamic memory allocation (GDL::vector instead of std::array) if TCol and TRow are -1.~~ POSTPONED - not necessary yet (and might not be in the future) ---> Use of templates reduce the need for dynamically sized matices.
- [x] Give AVX a second chance for dynamic allocations since I can use custom allocators for the 32 byte alignment. Additional note: Static versions should work too using the alignas(32) flag in the class definition.
- [x] rename static version to Mat and a furute dynamic version should be named MatX
- [x] ~~try initialization to zero with template meta programming the same way as the registers in the multiplication were initialized.~~ Tried it. No performance gain but a lot more code.
- [x] Add `operator +`
- [x] Add `operator ==`
- [x] Replace matrix comparison function with `operator ==` in MatX test
- [x] Test and benchmark `operator +`
- [x] Test and benchmark `operator ==` 
- [x] ~~Add a function that checks if all unused register values are 0! Add a `DEV_EXCEPTION` to every function that operates also on unused registers~~ Th unused register values should never effect any result. Hence it does not matter which value is stored inside them
- [x] Add `[[nodiscard]]` where appropriate
- [x] ~~Check if the multiplication helper function which creates the array of single value registers can be made faster by using shuffles/unpacks~~ Performance hit when compiled with clang. Gcc is just a little bit faster with shuffles
- [x] rename remove the X (matX) from test and benchmark names

**Mat4**
- [x] mat4 value constructor should take values in column major ordering to be consistent with matX. MatX can be initialized a lot faster by column major ordering, cause it stores its data in column major order and does not have to reorder the data provided by the constructor. 
- [x] Test operator()

**General**
- [x] remove all operator [] calls on register types (not portable, probably slow). Replace with an own sse getter function. Have a look at Cry Engine approach posted here ---> https://www.reddit.com/r/cpp/comments/7qj001/extract_elements_from_m128/ 
- [x] check for faster ways to initialize a register with a single value from another register ---> shuffles ---> shuffles are slower than storing the whole data in an array  an use _mm_set1 to create the new registers 
- [x] ~~make the alignment check in the ctors a free function (maybe in the sse helper header) which is called by the ctors. This way it is enough to check only once if the throw on misalignment is triggered in a test.~~
- [x] move special matrices (Rotation, Perspective) to own header - don't forget to remove cmath from the mat4Simd header in this case
- [x] replace SIMD with SSE in matrix names (and corresponding file names)

**Links**
- https://www.inf.ethz.ch/personal/markusp/teaching/263-2300-ETH-spring11/slides/class17.pdf

**Comment by vhirtham** [comment]
Finished with commit 0cbc3d25037a80741d9872688e6e6a7b92c02b01
